eclipse/sw360
# feat(rest): Added rest apis for create, update, delete vulnerability and create, delete vulnerability release relation

> Added rest apis for create, update, delete vulnerability.
>   * create, delete vulnerability release relation

> * Which issue is this pull request belonging to and how is it solving it? (*#1258*)
> * Did you add or update any new dependencies that are required for your change? - No

Issue: closes #1258 

### How To Test? - Refer Documentation
-  **3.9.3. Create a vulnerability**
     - Doc - http://localhost:8080/resource/docs/api-guide.html#resources-vulnerabilitie-create
 ```
 curl 'https://sw360.org/api/vulnerabilities' -i -X POST -H 'Content-Type: application/hal+json'
  -H 'Authorization: Bearer eyJhbGciOi*******' -H 'Accept: application/hal+json' 
  -d '{
  "externalId" : "987",
  "title" : "Title of vulnerability 12345",
  "description" : "Description of vulnerability",
  "publishDate" : "01-01-2018",
  "lastExternalUpdate" : "last-update-id-123",
  "priority" : "high",
  "priorityText" : "Priority Text",
  "action" : "new",
  "impact" : {
    "Key 1" : "Value 1",
    "Key 2" : "Value 2"
  },
  "legalNotice" : "Legal Notice",
  "assignedExtComponentIds" : [ "comp-id-2", "comp-id-1" ],
  "cveReferences" : [ {
    "year" : "2021",
    "number" : "4321"
  }, {
    "year" : "1992",
    "number" : "1234"
  } ],
  "vendorAdvisories" : [ {
    "vendor" : "vendor1",
    "name" : "name1",
    "url" : "http://url1"
  }, {
    "vendor" : "vendor2",
    "name" : "name2",
    "url" : "http://url2"
  } ],
  "extendedDescription" : "Extended Description",
  "references" : [ "reference-1", "reference-2" ],
  "cvss" : 3.0,
  "cvssTime" : "2019-01-01",
  "vulnerableConfiguration" : {
    "Key 1" : "Value 1",
    "Key 2" : "Value 2"
  },
  "access" : {
    "Key 1" : "Value 1",
    "Key 2" : "Value 2"
  },
  "cwe" : "common weakness enumeration",
  "cveFurtherMetaDataPerSource" : {
    "Cve meta data" : {
      "Key 1" : "Value 1",
      "Key 2" : "Value 2"
    }
  }
}'
```

![image](https://user-images.githubusercontent.com/56534710/121676364-29503f80-cad2-11eb-8182-7565a99afb17.png)

- 3.9.4. Update a vulnerability
   - Doc - http://localhost:8080/resource/docs/api-guide.html#resources-vulnerabilitie-update
```
curl 'https://sw360.org/api/vulnerabilities/98765' -i -X PATCH -H 'Content-Type: application/hal+json' -H 'Authorization: Bearer eyJhbGciOiJSUzI1N*********' -H 'Accept: application/hal+json' -d '{
  "externalId" : "98765",
  "title" : "Updated Title",
  "description" : "Description of vulnerability",
  "publishDate" : "01-01-2018",
  "lastExternalUpdate" : "last-update-id-123",
  "priority" : "high",
  "priorityText" : "Priority Text",
  "action" : "new",
  "impact" : {
    "Key 1" : "Value 1",
    "Key 2" : "Value 2"
  },
  "legalNotice" : "Legal Notice",
  "assignedExtComponentIds" : [ "comp-id-2", "comp-id-1" ],
  "cveReferences" : [ {
    "year" : "2021",
    "number" : "4321"
  }, {
    "year" : "1992",
    "number" : "1234"
  } ],
  "vendorAdvisories" : [ {
    "vendor" : "vendor1",
    "name" : "name1",
    "url" : "http://url1"
  }, {
    "vendor" : "vendor2",
    "name" : "name2",
    "url" : "http://url2"
  } ],
  "extendedDescription" : "Extended Description",
  "references" : [ "reference-1", "reference-2" ],
  "cvss" : 3.0,
  "cvssTime" : "2019-01-01",
  "vulnerableConfiguration" : {
    "Key 1" : "Value 1",
    "Key 2" : "Value 2"
  },
  "access" : {
    "Key 1" : "Value 1",
    "Key 2" : "Value 2"
  },
  "cwe" : "common weakness enumeration",
  "cveFurtherMetaDataPerSource" : {
    "Cve meta data" : {
      "Key 1" : "Value 1",
      "Key 2" : "Value 2"
    }
  }
}'
```

- **3.9.5. Delete a vulnerability**
   - Doc - http://localhost:8080/resource/docs/api-guide.html#resources-vulnerabilitie-delete
    ```
       curl 'https://sw360.org/api/vulnerabilities/123' -i -X DELETE 
       -H 'Authorization: Bearer ******' -H 'Accept: application/hal+json'
    ```

- **3.9.6. Create a release vulnerability relation**
  - Doc - http://localhost:8080/resource/docs/api-guide.html#resources-vulnerability-release-relation-create
```
curl 'https://sw360.org/api/vulnerabilities/98765/releaseVulnerabilityRelation' -i -X POST 
-H 'Content-Type: application/hal+json' -H 'Authorization: Bearer eyJhbGciOiJSUzI1NiI*******' -H 'Accept: application/hal+json'
 -d '{
  "releaseId" : "1234",
  "verificationStateInfo" : [ {
    "checkedOn" : "2021-06-17",
    "checkedBy" : "admin@sw360.org",
    "comment" : "Comment statement",
    "verificationState" : "CHECKED"
  } ],
  "matchedBy" : "matchedBy",
  "usedNeedle" : "usedNeedle"
}'
```

- **3.9.7. Delete a release vulnerability relation**
   - Doc - http://localhost:8080/resource/docs/api-guide.html#resources-vulnerability-release-relation-delete
```
curl 'https://sw360.org/api/vulnerabilities/98765/release/1234' -i -X DELETE 
-H 'Authorization: Bearer eyJhbGciOiJ***********' -H 'Accept: application/hal+json'
```

Note - It is required to test scenarios like view/edit project with linked release with linked vulnerability works after deleting Vulnerability or vulnerablity release relation. Same for release

> Have you implemented any additional tests? - No

### Checklist
Must:
